### PR TITLE
sql-parser: SSH TUNNEL connector parsing

### PR DESCRIPTION
**Pull request by mjibson**
Parsing for SSH TUNNEL connectors. No functionality yet.

### Motivation

  * This PR adds a known-desirable feature.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

File: src/storage/src/client/connections.rs [hunk 1]
```diff
@@ -713,3 +729,13 @@ impl Arbitrary for PostgresConnection {
...
     }
 }
+
+/// A connection to a SSH tunnel.
+#[derive(Arbitrary, Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct SshConnection {
+    pub host: String,
+    pub port: i32,
+    pub user: String,
+    pub public_key: String,
```
Comment (mjibson) [inline src/storage/src/client/connections.rs:739]: These keys could arguably be `Vec<u8>`, but it's also pretty reasonable to keep them as the human readable strings here and force the lower layers to convert those to the usable bytes.

File: src/storage/src/client/connections.rs [hunk 2]
```diff
@@ -713,3 +729,13 @@ impl Arbitrary for PostgresConnection {
...
+    pub private_key: String,
```
Comment (benesch) [inline src/storage/src/client/connections.rs:740]: `private_key` should be a secret!

```suggestion
    pub private_key: GlobalId,
```

File: src/adapter/src/catalog/builtin.rs [hunk 3]
```diff
@@ -1020,6 +1020,14 @@ pub static MZ_CONNECTIONS: Lazy<BuiltinTable> = Lazy::new(|| BuiltinTable {
         .with_column("name", ScalarType::String.nullable(false))
         .with_column("type", ScalarType::String.nullable(false)),
 });
+pub static MZ_SSH_TUNNEL_CONNECTORS: Lazy<BuiltinTable> = Lazy::new(|| BuiltinTable {
+    name: "mz_ssh_tunnel_connectors",
```
Comment (benesch) [inline src/adapter/src/catalog/builtin.rs:1024]:
```suggestion
pub static MZ_SSH_TUNNEL_CONNECTIONS: Lazy<BuiltinTable> = Lazy::new(|| BuiltinTable {
    name: "mz_ssh_tunnel_connections",
```

File: src/sql/src/plan/statement/ddl.rs [hunk 4]
```diff
@@ -3066,6 +3096,11 @@ pub fn plan_create_connection(
             let connection = mz_storage::client::connections::PostgresConnection::try_from(c)?;
             Connection::Postgres(connection)
         }
+        CreateConnection::Ssh { with_options } => {
```
Comment (benesch) [inline src/sql/src/plan/statement/ddl.rs:3099]:
```suggestion
        CreateConnection::Ssh { with_options } => {
            scx.require_unsafe_mode("CREATE CONNECTION ... SSH")?;
```

(for now)

**Review by benesch — APPROVED** [review]

**Comment by mjibson** [comment]
@pH14 this should be ready for storage to take over now?

**Comment by pH14** [comment]
Amazing, thank you @mjibson !

**Pull request merged**